File: backend/sarvam_translate.py
import requests
import os

SARVAM_API_KEY = os.getenv("SARVAM_API_KEY", "sk_y9798qpa_JJoiaSJY4GWdgVWW8Gs0LGN5")
SARVAM_URL = "https://api.sarvam.ai/translate"
MAX_CHARS = 1000

def sarvam_translate(text, target_language_code, source_language_code="auto"):
    headers = {
        "api-subscription-key": SARVAM_API_KEY,
        "Content-Type": "application/json"
    }
    # Split text into chunks of <= 1000 characters
    chunks = [text[i:i+MAX_CHARS] for i in range(0, len(text), MAX_CHARS)]
    translated_chunks = []
    for chunk in chunks:
        payload = {
            "input": chunk,
            "source_language_code": source_language_code,
            "target_language_code": target_language_code
        }
        response = requests.post(SARVAM_URL, headers=headers, json=payload)
        if response.status_code == 200:
            translated_chunks.append(response.json().get("translated_text"))
        else:
            raise Exception(f"Sarvam API error: {response.status_code} - {response.text}")
    return "".join(translated_chunks) 

File: react-frontend/src/utils/sarvamTranslate.js
const SARVAM_API_KEY = "sk_y9798qpa_JJoiaSJY4GWdgVWW8Gs0LGN5";
const SARVAM_URL = "https://api.sarvam.ai/translate";
const MAX_CHARS = 2000;

export async function sarvamTranslate(input, targetLang, sourceLang = "auto") {
  const normalizeCode = code => (code && code.includes('-') ? code.split('-')[0] : code);

  // Split input into chunks of <= MAX_CHARS
  const chunks = [];
  for (let i = 0; i < input.length; i += MAX_CHARS) {
    chunks.push(input.slice(i, i + MAX_CHARS));
  }

  let translated = "";
  for (const chunk of chunks) {
    // Try BCP-47 code first
    console.log(`[Sarvam Translate] Translating chunk: source=${sourceLang}, target=${targetLang}, length=${chunk.length}`);
    let response = await fetch(SARVAM_URL, {
      method: "POST",
      headers: {
        "api-subscription-key": SARVAM_API_KEY,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        input: chunk,
        source_language_code: sourceLang,
        target_language_code: targetLang
      })
    });

    // If 400, try fallback with normalized codes
    if (!response.ok && response.status === 400 && (targetLang.includes('-') || sourceLang.includes('-'))) {
      const normSource = normalizeCode(sourceLang);
      const normTarget = normalizeCode(targetLang);
      console.log(`[Sarvam Translate] Fallback translation: source=${normSource}, target=${normTarget}`);
      response = await fetch(SARVAM_URL, {
        method: "POST",
        headers: {
          "api-subscription-key": SARVAM_API_KEY,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          input: chunk,
          source_language_code: normSource,
          target_language_code: normTarget
        })
      });
    }

    if (!response.ok) throw new Error("Sarvam Translate failed");
    const data = await response.json();
    translated += data.translated_text || "";
  }
  return { translated_text: translated };
} 